Error handling in nlp.pipe

svlandeg opened the pull request:
This seems a frequently asked feature - e.g. https://github.com/explosion/spaCy/issues/6630 and https://github.com/explosion/spaCy/issues/4847, and I've been wanting it myself too ;-)

## Description
With the `Pipe` / `TrainablePipe` API revamped for spaCy 3, I thought it would make sense to also add an error handler to `nlp.pipe` and the `pipe` methods of trainable pipeline components.

The default error handler changes nothing: it just reraises the same exception. But the user has an option of specifying its own error handling function that can ignore the batch, inspect the batch, warn, ... 

## Open question
Which arguments should the error handler function take? Currently it takes the component's name, the offending batch of documents, and the original error that was thrown.

### Types of change
enhancement, UX

## Checklist
- [x] I have submitted the spaCy Contributor Agreement.
- [x] I ran the tests, and all new and existing tests passed.
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

## TODO
- [ ] In multi-processing context

honnibal commented:
This is a really cool idea! I'll avoid merging it just yet so we can think about the open questions, but I'm definitely in favour of this.

honnibal commented:
> Which arguments should the error handler function take? Currently it takes the component's name, the offending batch of documents, and the original error that was thrown.

May as well pass the component itself, right?

honnibal commented:
Do we have to have this in the components?

I'd rather have it just on the `Language.pipe`. It seems like the sort of argument people will always be forgetting to add to their components' `pipe()` methods, causing errors at unexpected times.

I don't know how to make the error handling work in `Language` with the generator trickery though. But I do get worried at this change to the `pipe` signature.

honnibal commented:
How come we have so many of these `pipe` implementations, anyway? Aren't there a bunch that could be using `TrainablePipe.pipe`?

svlandeg commented:
> Edit: There's really no way to make it work in `Language.pipe` without passing it into the generators themselves, is there. 

I don't directly see how. Alternatively, we could change the API of the components `pipe` methods entirely and make them take in a batch instead of a generator, and do the batching (only) in `Language`, but that's equally breaking and will cause a bit of headache as well.

honnibal commented:
The `component.pipe()` thing has more disadvantages than I realised when I did it, but it also has advantages, and I think we should accept that it's just How Things Are.

How about this, to avoid changing the arguments of `component.pipe()`:

* Add new methods `Pipe.set_error_handler` and `Pipe.get_error_handler`
* Add a new method `set_error_handler` to `Language`. This will go over the pipeline and set the error handler on components, _if_ they have a `set_error_handler` method. Alternatively, you can set different error handlers to specific components.
* Inside the `.pipe` and `.__call__` methods, we try to get the error handler, or use the default one if we don't have one. Again we don't assume the API is there, we have to do `hasattr` checks etc (gross, I know...).

I think carrying the `error_handler` as state instead of an argument is a bit nicer, because it lets you set different error handlers on different components. It also doesn't really depend on the data you want to pipe through, and it's something I can imagine setting in the config, which wouldn't work if it's an argument of `Language.pipe`.

svlandeg commented:
I've been thinking about this some more and agree that keeping this as a state instead of an additional argument probably makes more sense. I changed it such that the `Language` object (also) keeps track of a default error handler, so that it can do something useful also for components that don't do it themselves (like simple functions).

I'm still a little unsure about the current implementation.  `TrainablePipe` now takes care of error handling in its `pipe` method (and I removed the redundant overriding duplicates) but then what about the `__call__` methods? It feels like it would be weird not to have the same error handling when you do `proc(doc)`, especially considering the error handler is now saved on the `proc` object. But if you want to make this consistent, you end up with large `try-except` blocks in other `__call__` methods like the one in `AttributeRuler`, and I really really detest those. Then again we're not catching and ignoring, we're just giving the user the option to do whatever, which is nice I think. It just looks really awful.

honnibal commented:
> But if you want to make this consistent, you end up with large try-except blocks in other __call__ methods like the one in AttributeRuler, and I really really detest those. Then again we're not catching and ignoring, we're just giving the user the option to do whatever, which is nice I think. It just looks really awful.

Ugh, yeah. We can at least refactor these. Even just moving the body to a `_call` method is a more than cosmetic improvement, because if you subclass, you can change the `_call` and get the error handling from the parent.

I find the `.predict()` and `set_annotations()` division appealing. If we find `predict()` misleading for a "ruler", we could call it `match()`?

svlandeg commented:
> (I sense the growth of another base type here, something like `MatcherPipe`, but let's not get into that.)

I shall restrain myself for now ;-)
